### PR TITLE
Add buffer-pool command to calculate ribps

### LINKED ISSUE #180: Add command for calculating recommended innodb buffer pool size of a site
Take into account environment for the 'factor', then run the calculation command against the site's db.

### PR DESCRIPTION
**Pull request by joshbetz**
https://dba.stackexchange.com/a/27341

Fixes #180

### CODE REVIEW

File: src/lib/db.js [hunk 1]
```diff
@@ -96,3 +96,16 @@ export function getCLI( site, callback ) {
...
+export function query( site, query, callback ) {
+	getConnection( site, ( err, args ) => {
+		if ( err ) {
+			return callback( err );
+		}
+
+		args.push( '--safe-updates' );
+		args.push( '-e', query );
+
+		spawn( 'mysql', args, { stdio: 'inherit' });
```
Comment (nickdaugherty) [inline src/lib/db.js]: Will this exit after running the query and printing the results, or keep a MySQL CLI instance open?

File: src/bin/vip-db.js [hunk 2]
```diff
@@ -63,6 +63,31 @@ program
...
+	.command( 'buffer-pool <site>' )
+	.action( ( site, options ) => {
+		try {
+			which.sync( 'mysql' );
+		} catch ( e ) {
+			return console.error( 'MySQL client is required and not installed.' );
+		}
+
+		utils.findSite( site, ( err, site ) => {
+			let factor = site.environment === 'production' ? 1.6 : 1;
```
Comment (nickdaugherty) [inline src/bin/vip-db.js]: Given the System's team concerns around using too much memory, should we start with a more conservative set of factors?

**Comment by joshbetz** [comment]
Example output:

```
$ vip db buffer-pool v2.joshbetz.com
+------------+-------------+-------------+-----------+
| data_bytes | index_bytes | total_bytes | innodb_mb |
+------------+-------------+-------------+-----------+
|   17793024 |    14778368 |    32571392 |        32 |
+------------+-------------+-------------+-----------+
```

File: src/lib/db.js (hunk 1 above)
Comment (joshbetz) [inline src/lib/db.js]: The `-e` parameter runs a query and exits. Added example output below.

File: src/bin/vip-db.js (hunk 2 above)
Comment (joshbetz) [inline src/bin/vip-db.js]: Makes sense to me. What are you thinking? Maybe 1.2?

Comment (nickdaugherty) [inline src/bin/vip-db.js]: Ya that sounds decent, then maybe on non-production, we go down to 0.75?

**Comment by nickdaugherty** [comment]
> I'm wondering whether this would be easier to read if everything was in MB.

That's a good idea, especially since we control the InnoDB settings as MB in meta.

**Comment by nickdaugherty** [comment]
> I think the buffer pool always needs to be at least as big as index_bytes.

Depends who you ask - from Barry's comment linked above:

> IMHO it’s not practical or required to have the entire dataset stored in RAM, which is what the calculation does.

**Comment by joshbetz** [comment]
@nickdaugherty Right, not the entire data set, but I think the indexes should be?

**Comment by mjangda** [comment]
> Barry has been pushing for looking at the ratio between read requests and reads as a tuning metric, are there plans to add that calculation too?

I believe David had noted that he was going to add some tracking for this?

File: src/bin/vip-db.js [hunk 3]
```diff
@@ -63,6 +63,41 @@ program
...
+	.option( '-f, --factor <factor>', 'Database growth factor', 0, parseFloat )
+	.action( ( site, options ) => {
+		try {
+			which.sync( 'mysql' );
+		} catch ( e ) {
+			return console.error( 'MySQL client is required and not installed.' );
+		}
+
+		utils.findSite( site, ( err, site ) => {
+			var factor;
```
Comment (mjangda) [inline src/bin/vip-db.js]: `let`

File: src/bin/vip-db.js [hunk 4]
```diff
@@ -63,6 +63,41 @@ program
...
+	.command( 'buffer-pool <site>' )
+	.option( '-f, --factor <factor>', 'Database growth factor', 0, parseFloat )
+	.action( ( site, options ) => {
+		try {
+			which.sync( 'mysql' );
+		} catch ( e ) {
+			return console.error( 'MySQL client is required and not installed.' );
+		}
+
+		utils.findSite( site, ( err, site ) => {
```
Comment (mjangda) [inline src/bin/vip-db.js]: Can we add headers similar to what we have for the `db` command?

File: src/bin/vip-db.js [hunk 5]
```diff
@@ -63,6 +63,41 @@ program
...
+			}
+
+			let query = `SELECT
+				CEILING(SUM(data_length)/POWER(1024,2)) data_mb,
+				CEILING(SUM(index_length)/POWER(1024,2)) index_mb,
+				CEILING(SUM(data_length+index_length)/POWER(1024,2)) total_mb,
+				CEILING(SUM(data_length+index_length)*${factor}/POWER(1024,2)) innodb_mb
+			FROM information_schema.tables WHERE engine='InnoDB'`;
+
+			db.query( site, query, err => {
```
Comment (mjangda) [inline src/bin/vip-db.js]: Might be nice to note the factor that is being used.

File: src/bin/vip-db.js (hunk 3 above)
Comment (joshbetz) [inline src/bin/vip-db.js]: Good call. I've been bitten by scoping issues related to `let` in the past, so I have a bad habit of defaulting to `var`.

**Comment by joshbetz** [comment]
```
$ vip db buffer-pool v2.joshbetz.com
-- Site: 59
-- Domain: v2.joshbetz.com
-- Environment: testing
-- Growth Factor: 0.75
+---------+----------+----------+-----------+
| data_mb | index_mb | total_mb | innodb_mb |
+---------+----------+----------+-----------+
|      17 |       15 |       32 |        24 |
+---------+----------+----------+-----------+
```

File: src/bin/vip-db.js (hunk 4 above)
Comment (joshbetz) [inline src/bin/vip-db.js]: Added

File: src/bin/vip-db.js (hunk 5 above)
Comment (joshbetz) [inline src/bin/vip-db.js]: Added

**Pull request merged**